### PR TITLE
add metax guide

### PR DESCRIPTION
**Pull request by offgridauthor**

### CODE REVIEW

File: docs/modules/ROOT/pages/guide-metatx.adoc [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+
+[[demo-app-overview]]
+== Demo App Overview
+
+You can view the live https://defender-metatx-workshop-demo.openzeppelin.com/[demo dapp] here. It accepts registrations directly if the user has the available funds to pay for the transaction, otherwise the data is sent as a meta-transaction.
+
+In the example code, the functionality of the https://github.com/OpenZeppelin/workshops/blob/master/25-defender-metatx-api/contracts/SimpleRegistry.sol[`SimpleRegistry` contract] is to take a string and store it. The contract's https://github.com/OpenZeppelin/workshops/blob/master/25-defender-metatx-api/contracts/Registry.sol[metatransaction implementation] achieves the same result by decoupling the signer from the sender of the transaction.
+
+When comparing the code, note the metatransaction's use of `_msgSender()` as opposed to the SimpleRegistry's use of `msg.sender`. By extending from `ERC2771Context` and `MinimalForwarder`, the contract becomes metatransaction capable. (Note that all OpenZeppelin contracts are compatable with the use of `_msgSender()`.)
```
Comment (mverzilli) [inline docs/modules/ROOT/pages/guide-metatx.adoc:16]:
```suggestion
When comparing the code, note the metatransaction's use of `_msgSender()` as opposed to the SimpleRegistry's use of `msg.sender`. By extending from `ERC2771Context` and `MinimalForwarder`, the contract becomes metatransaction capable. (Note that all OpenZeppelin contracts are compatible with the use of `_msgSender()`.)
```

**Review by mverzilli — APPROVED** [review]

File: docs/modules/ROOT/pages/guide-metatx.adoc [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+[[create-relayer]]
+== Create Relayer
+
+First, install the necessary package:
+
+`npm install defender-relay-client`
+
+Create a script such as the following, supplying your Defender API key in your .env file and adjusting variable names as appropriate.
+
```
Comment (spalladino) [inline docs/modules/ROOT/pages/guide-metatx.adoc:47]: I'd include the script in the repo, so they don't have to copy-paste it from here. We can include in the guide only the relevant parts (ie the lines that actually create the relayer)

File: docs/modules/ROOT/pages/guide-metatx.adoc [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+  
+  // store relayer info in file
+  writeFileSync('relay.json', JSON.stringify({
+    relayer
+  }, null, 2));
+  console.log('Relayer ID: ', relayer);
+
+  // create and save the api key to .env - needed for sending tx
+  const {relay} = JSON.parse(readFileSync('relay.json'))
+  const secretKey = await relayClient.createKey(relayer.relayerId);
```
Comment (spalladino) [inline docs/modules/ROOT/pages/guide-metatx.adoc:73]: We already have the relayerId, no need to re-read relay.json

File: docs/modules/ROOT/pages/guide-metatx.adoc [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+  writeFileSync('.env.relay', JSON.stringify({secretKey}, null, 2));
+}
+
+run().catch((error) => {
+  console.error(error);
+  process.exitCode = 1;
+});
+```
+
+Run with `yarn create-relay`
```
Comment (spalladino) [inline docs/modules/ROOT/pages/guide-metatx.adoc:83]: Let's clarify that this can also be done in the UI by manually creating a relay, and link to the docs

File: docs/modules/ROOT/pages/guide-metatx.adoc [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+  // store relayer info in file
+  writeFileSync('relay.json', JSON.stringify({
+    relayer
+  }, null, 2));
+  console.log('Relayer ID: ', relayer);
+
+  // create and save the api key to .env - needed for sending tx
+  const {relay} = JSON.parse(readFileSync('relay.json'))
+  const secretKey = await relayClient.createKey(relayer.relayerId);
+  writeFileSync('.env.relay', JSON.stringify({secretKey}, null, 2));
```
Comment (spalladino) [inline docs/modules/ROOT/pages/guide-metatx.adoc:74]: Nit: the convention for multiple env files is to separate them by environment, so this should rather be `.env.local` (see [here](https://www.prisma.io/docs/guides/development-environment/environment-variables/using-multiple-env-files)). OTOH, dotenv advises against this (see FAQ [here](https://www.npmjs.com/package/dotenv)). Also, env files should be in bash format (RELAY_KEY=foobar), not json.

Bottom line: I'd suggest you either programmatically add the relayer key to .env (`appendFileSync('.env', `RELAY_KEY=${secretKey}`), so you automatically get it in the process.env when you run require `dotenv`, or store it in a separate `key.json` file. Whichever of the two you pick, the file must be added to the gitginore.

File: docs/modules/ROOT/pages/guide-metatx.adoc [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+    names[owner] = name;
+    emit Registered(owner, name);
+  }
+}
+```
+
+Run `npx hardhat compile` to get the code ready for deployment.
+
+[[deploy-using-relayer]]
+== Deploy Using Relayer
```
Comment (spalladino) [inline docs/modules/ROOT/pages/guide-metatx.adoc:124]: This is awesome just in itself - we should make a guide in the forum about it. We'll write a hardhat plugin for this eventually, but for now, this is great!!

File: docs/modules/ROOT/pages/guide-metatx.adoc [hunk 7]
```diff
@@ -0,0 +1,349 @@
...
+    .catch(error => { console.error(error); process.exit(1); });
+}
+
+```
+
+This code pulls the Relayer's credentials from the local file along with the artifacts for the `Registry` and `MinimalForwarder` contracts and uses ethers.js to deploy. The relevant addresses of these contracts are saved to the local file `deploy.json`.
+
+In the demo app, run this script with `yarn deploy`
+
+Note: As the owners of the Relayer, are the paying for the gas, so it is necessary to keep the Relayer funded sufficiently to cover future transactions. For Defender users on testnets, Relayer balances are auto-topped as necessary to ensure transactions go through.
```
Comment (spalladino) [inline docs/modules/ROOT/pages/guide-metatx.adoc:167]: Note that this only applies to Goerli as of today

File: docs/modules/ROOT/pages/guide-metatx.adoc [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+
+if (require.main === module) {
+  main().then(() => process.exit(0))
+    .catch(error => { console.error(error); process.exit(1); });
+}
+```
+
+Next, create the Autotask itself. The main idea here is to supply the necessary logic for telling the Relayer to send a transaction to the forwarder contract, supplying the signer's address. The Autotask will get triggered by a call to its webhook from the dapp.
+
+Due to the tight relationship between components, the Relayer credentials are securely available to the Autotask simply by instantiating a new provider and signer. Note that the address of the `MinimalForwarder` and the `Registry` contract must be supplied.
```
Comment (spalladino) [inline docs/modules/ROOT/pages/guide-metatx.adoc:211]: Good point - why are we storing the relayer secret key to the local env file then?

File: docs/modules/ROOT/pages/guide-metatx.adoc [hunk 9]
```diff
@@ -0,0 +1,349 @@
...
+
+This code pulls the Relayer's credentials from the local file along with the artifacts for the `Registry` and `MinimalForwarder` contracts and uses ethers.js to deploy. The relevant addresses of these contracts are saved to the local file `deploy.json`.
+
+In the demo app, run this script with `yarn deploy`
+
+Note: As the owners of the Relayer, are the paying for the gas, so it is necessary to keep the Relayer funded sufficiently to cover future transactions. For Defender users on testnets, Relayer balances are auto-topped as necessary to ensure transactions go through.
+
+
+[[create-autotask]]
+== Create Autotask via API
```
Comment (spalladino) [inline docs/modules/ROOT/pages/guide-metatx.adoc:171]: We very frequently see users that try to expose their relayer directly to the frontend, bypassing the autotask. We should explain here why it's important to have the autotask: so you can decide whether to pay for the request or not. And that if you expose the relayer directly, anyone can just get your relayer keys and use the relayer to send whatever tx they want.

**Review by spalladino — COMMENTED** [review]
Awesome work @offgridauthor!! My main comment would be to include these scripts on the sample repo, so the user doesn't have to copy-paste them around.
